pandas/core/reshape/merge.py

```diff
@@ -982,6 +985,12 @@ def _maybe_add_join_keys(
                     key_col = Index(lvals).where(~mask_left, rvals)
                     result_dtype = find_common_type([lvals.dtype, rvals.dtype])
 
+                if (self.left_index and not self.right_index) or (
+                    self.right_index and not self.left_index
+                ):
+                    if key_col.equals(result.index):
```

Comment:
Might be misreading this but what is the point of this loop? Only ever continues?

Reply:
this would hit continue, only when it's merging on a col and an index and index already contains correct values